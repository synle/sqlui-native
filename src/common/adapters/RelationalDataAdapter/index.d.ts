import BaseDataAdapter from 'src/common/adapters/BaseDataAdapter/index';
import IDataAdapter from 'src/common/adapters/IDataAdapter';
import { SqluiCore } from 'typings';
export default class RelationalDataAdapter extends BaseDataAdapter implements IDataAdapter {
    dialect?: SqluiCore.Dialect;
    constructor(connectionOption: string);
    private getConnection;
    authenticate(): Promise<void>;
    getDatabases(): Promise<SqluiCore.DatabaseMetaData[]>;
    getTables(database?: string): Promise<SqluiCore.TableMetaData[]>;
    getColumns(table: string, database?: string): Promise<SqluiCore.ColumnMetaData[]>;
    execute(sql: string, database?: string): Promise<SqluiCore.Result>;
    private _execute;
}
