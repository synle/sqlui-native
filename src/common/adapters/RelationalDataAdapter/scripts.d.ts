import BaseDataScript from 'src/common/adapters/BaseDataAdapter/scripts';
import { SqlAction } from 'typings';
export declare function getSelectAllColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getSelectCount(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getSelectSpecificColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getSelectDistinctValues(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getInsert(input: SqlAction.TableInput, value?: Record<string, any>): SqlAction.Output | undefined;
export declare function getBulkInsert(input: SqlAction.TableInput, rows?: Record<string, any>[]): SqlAction.Output | undefined;
export declare function getUpdate(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getUpdateWithValues(input: SqlAction.TableInput, value: Record<string, any>, conditions: Record<string, any>): SqlAction.Output | undefined;
export declare function getDelete(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getCreateTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropTable(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getAddColumn(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropColumns(input: SqlAction.TableInput): SqlAction.Output | undefined;
export declare function getDropDatabase(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getCreateDatabase(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getCreateSampleTable(input: SqlAction.DatabaseInput): SqlAction.Output | undefined;
export declare function getCreateConnectionDatabase(input: SqlAction.ConnectionInput): SqlAction.Output | undefined;
export declare class ConcreteDataScripts extends BaseDataScript {
    dialects: string[];
    getIsTableIdRequiredForQuery(): boolean;
    getSyntaxMode(): string;
    supportMigration(): boolean;
    supportCreateRecordForm(): boolean;
    supportEditRecordForm(): boolean;
    supportVisualization(): boolean;
    getDialectIcon(dialect: any): string;
    getTableScripts(): (typeof getInsert)[];
    getDatabaseScripts(): (typeof getDropDatabase)[];
    getConnectionScripts(): (typeof getCreateConnectionDatabase)[];
    getSampleConnectionString(dialect: any): "" | "mssql://sa:password123!@localhost:1433" | "postgres://postgres:password@localhost:5432" | "sqlite://test-db.sqlite" | "mariadb://root:password@localhost:3306" | "mysql://root:password@localhost:3306";
    getSampleSelectQuery(tableActionInput: any): SqlAction.Output;
    getCodeSnippet(connection: any, query: any, language: any): string;
}
declare const _default: ConcreteDataScripts;
export default _default;
