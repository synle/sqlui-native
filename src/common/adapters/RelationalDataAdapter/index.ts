import qs from 'qs';
import { Sequelize } from 'sequelize';
import BaseDataAdapter from 'src/common/adapters/BaseDataAdapter/index';
import IDataAdapter from 'src/common/adapters/IDataAdapter';
import { SqluiCore } from 'typings';

const DEFAULT_SEQUELIZE_OPTION = {
  logging: false,
  dialectOptions: {
    multipleStatements: true,
  },
};

/**
 * mostly adapter for sequelize
 * https://sequelize.org/master/class/lib/dialects/abstract/query-interface.js~QueryInterface.html
 */
export default class RelationalDataAdapter extends BaseDataAdapter implements IDataAdapter {
  dialect?: SqluiCore.Dialect;
  private sequelizes: Record<string, Sequelize> = {};

  constructor(connectionOption: string) {
    super(connectionOption);

    // since mariadb and mysql are fully compatible, let's use the same data
    // save the connection string
    this.connectionOption = this.connectionOption.replace('mariadb://', 'mysql://');

    // TODO: we don't support sslmode, this will attempt to override the option
    this.connectionOption = this.connectionOption.replace('sslmode=require', 'sslmode=no-verify');

    // save the root connection
    this.sequelizes[''] = new Sequelize(this.connectionOption);
  }

  private getConnection(database: string = ''): Sequelize {
    if (!this.sequelizes[database]) {
      let connectionUrl: string;
      let connectionPropOptions : any = {...DEFAULT_SEQUELIZE_OPTION};

      switch (this.dialect) {
        case 'sqlite':
          database = '';

          // special handling for sqlite path
          let sqliteStorageOption = this.connectionOption
            .replace('sqlite://', '')
            .replace(/\\/g, '/'); // uses :memory: for in memory

          connectionUrl = `sqlite://`
          connectionPropOptions = {...connectionPropOptions, ...{storage: sqliteStorageOption,}} // applicable for sqlite}}
          break;

        default:
          //@ts-ignore
          const { scheme, username, password, hosts, options } =
            BaseDataAdapter.getConnectionParameters(this.connectionOption);

          connectionUrl = `${scheme}://`;
          if (username && password) {
            connectionUrl += `${encodeURIComponent(username)}:${encodeURIComponent(password)}`;
          }

          const [{ host, port }] = hosts;
          connectionUrl += `@${host}:${port}`;

          if (database) {
            connectionUrl += `/${database}`;
          }

          if (options) {
            connectionUrl += `?${qs.stringify(options)}`;
          }
          break;
      }

      this.sequelizes[database] = new Sequelize(connectionUrl, connectionPropOptions);
    }

    return this.sequelizes[database];
  }

  async authenticate() {
    return this.getConnection().authenticate();
  }

  async getDatabases(): Promise<SqluiCore.DatabaseMetaData[]> {
    let sql;

    switch (this.dialect) {
      case 'mssql':
        sql = `SELECT name AS 'database' FROM sys.databases`;
        break;
      case 'postgres':
        sql = `SELECT datname AS database FROM pg_database`;
        break;
      case 'sqlite':
        // because SQLITE doesn't have a concept of database
        // so we will hard code this as sqlite here
        // so that the ui will show up
        return [
          {
            name: 'Sqlite',
            tables: [], // TODO: will remove this entirely
          },
        ];
        break;
      case 'mariadb':
      case 'mysql':
        sql = `show databases`;
        break;
    }

    if (!sql) {
      return [];
    }

    const [data] = await this._execute(sql);

    return data
      .map(
        (row: any) =>
          row.Database || // mssql and mysql
          row.database, // postgres
      )
      .filter((db) => db)
      .map((name) => ({
        name,
        tables: [], // TODO: will remove this entirely
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getTables(database?: string): Promise<SqluiCore.TableMetaData[]> {
    // https://github.com/knex/knex/issues/360
    // PostgreSQL: SELECT tablename FROM pg_tables WHERE schemaname='public''
    // MySQL: SELECT TABLE_SCHEMA FROM information_schema.tables GROUP BY TABLE_SCHEMA;
    // SQLite3: SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE '%sqlite%';
    let sql;
    let data;

    switch (this.dialect) {
      case 'mssql':
        sql = `SELECT name AS 'tablename' FROM SYSOBJECTS WHERE xtype = 'U' ORDER BY tablename`;
        break;
      case 'postgres':
        sql = `SELECT tablename FROM pg_tables WHERE schemaname='public' ORDER BY tablename`;
        break;
      default:
      case 'sqlite':
        sql = `SELECT name AS tablename FROM sqlite_master WHERE type='table' AND name NOT LIKE '%sqlite%' ORDER BY tablename`;
        break;
      case 'mariadb':
      case 'mysql':
        sql = `SELECT TABLE_NAME as tablename FROM information_schema.tables WHERE TABLE_SCHEMA = '${database}' ORDER BY tablename`;
        break;
    }

    if (!sql) {
      return [];
    }

    [data] = await this._execute(sql, database);

    return data
      .map((row: any) => row.tablename)
      .filter((db) => db)
      .map((name) => ({
        name,
        columns: [], // TODO: will remove this entirely
      }))
      .sort((a, b) => (a.name || '').localeCompare(b.name || ''));
  }

  async getColumns(table: string, database?: string): Promise<SqluiCore.ColumnMetaData[]> {
    switch (this.dialect) {
      case 'mssql':
      case 'postgres':
      case 'sqlite':
      case 'mariadb':
      case 'mysql':
      default:
        // first get all the columns
        const columns: SqluiCore.ColumnMetaData[] = [];
        try {
          const columnMap = await this.getConnection(database)
            .getQueryInterface()
            .describeTable(table);

          for (const columnName of Object.keys(columnMap)) {
            columns.push({
              name: columnName,
              ...columnMap[columnName],
            });
          }
        } catch (err) {}

        // then see if we attempt to add additional foreignKey constraint
        try {
          const foreignKeyReferences = (await this.getConnection(database)
            .getQueryInterface()
            .getForeignKeyReferencesForTable(table)) as any[];

          for (const foreignKeyReference of foreignKeyReferences) {
            const fromTableName = foreignKeyReference.tableName;
            const fromColumnName = foreignKeyReference.columnName;
            const toTableName = foreignKeyReference.referencedTableName;
            const toColumnName = foreignKeyReference.referencedColumnName;

            if (fromTableName === table) {
              const targetColumn = columns.find((column) => column.name === fromColumnName);

              if (targetColumn) {
                targetColumn.kind = 'foreign_key';
                targetColumn.referencedTableName = toTableName;
                targetColumn.referencedColumnName = toColumnName;
              }
            }
          }
        } catch (err) {}

        return columns;
    }
  }

  async execute(sql: string, database?: string): Promise<SqluiCore.Result> {
    // https://sequelize.org/master/manual/raw-queries.html
    //@ts-ignore
    try {
      let [raw, meta] = await this.getConnection(database).query(sql, {
        raw: true,
        plain: false,
      });

      let rawToUse: any | undefined = raw;
      let metaToUse: any | undefined = meta;
      let affectedRows;

      switch (this.dialect) {
        case 'mssql':
          affectedRows = metaToUse;
          break;
        case 'postgres':
          if (metaToUse.rowCount >= 0) {
            affectedRows = metaToUse.rowCount;
          }
          // Postgres returns a lot of redundant data, best to remove it to save space...
          metaToUse = undefined;
          break;
        case 'sqlite':
          if (metaToUse.changes >= 0) {
            affectedRows = metaToUse.changes;
          } else {
            metaToUse = undefined;
          }
          break;
        case 'mariadb':
        case 'mysql':
          if (metaToUse?.affectedRows) {
            // these are likely insert / update calls
            // these don't need raw data
            rawToUse = undefined;
            affectedRows = metaToUse?.affectedRows;
          } else {
            // these are select calls, we don't need the meta
            metaToUse = undefined;
          }
        default:
          break;
      }

      return {
        ok: true,
        raw: rawToUse,
        meta: metaToUse,
        affectedRows,
      };
    } catch (error) {
      return {
        ok: false,
        error,
      };
    }
  }

  private async _execute(
    sql: string,
    database?: string,
  ): Promise<[SqluiCore.RawData, SqluiCore.MetaData]> {
    // https://sequelize.org/master/manual/raw-queries.html
    //@ts-ignore
    return this.getConnection(database).query(sql, {
      raw: true,
      plain: false,
    });
  }
}
