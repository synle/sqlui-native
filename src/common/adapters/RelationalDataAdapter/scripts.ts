import BaseDataScript, { getDivider } from 'src/common/adapters/BaseDataAdapter/scripts';
import { escapeSQLValue, isValueNumber } from 'src/frontend/utils/formatter';
import { SqlAction, SqluiCore } from 'typings';

const formatter = 'sql';

export function getSelectAllColumns(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Select All Columns`;

  switch (input.dialect) {
    case 'mssql':
      return {
        label,
        formatter,
        query: `SELECT TOP ${input.querySize} *
                FROM ${input.tableId}`,
      };
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `SELECT *
                FROM ${input.tableId}
                LIMIT ${input.querySize}`,
      };
  }
}

export function getSelectCount(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Select Count`;

  if (!input.columns) {
    return undefined;
  }

  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join(' AND \n');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `SELECT COUNT(*)
                FROM ${input.tableId}
                WHERE ${whereColumnString}`,
      };
  }
}

export function getSelectSpecificColumns(
  input: SqlAction.TableInput,
): SqlAction.Output | undefined {
  const label = `Select Specific Columns`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = `\n` + input.columns.map((col) => `  ${col.name}`).join(',\n');
  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join('\n AND ');

  switch (input.dialect) {
    case 'mssql':
      return {
        label,
        formatter,
        query: `SELECT TOP ${input.querySize} ${columnString}
                FROM ${input.tableId}
                WHERE ${whereColumnString}`,
      };
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `SELECT ${columnString}
                FROM ${input.tableId}
                WHERE ${whereColumnString}
                LIMIT ${input.querySize}`,
      };
  }
}

export function getSelectDistinctValues(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Select Distinct`;

  if (!input.columns) {
    return undefined;
  }

  const columns = input.columns || [];

  const distinctColumn = columns.filter((col) => !col.primaryKey)?.[0]?.name || 'some_field';
  const whereColumnString = columns.map((col) => `${col.name} = ''`).join('\n AND ');

  switch (input.dialect) {
    case 'mssql':
      return {
        label,
        formatter,
        query: `SELECT DISTINCT TOP ${input.querySize} ${distinctColumn}
                FROM ${input.tableId}
                WHERE ${whereColumnString}`,
      };
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `SELECT DISTINCT ${distinctColumn}
                FROM ${input.tableId}
                WHERE ${whereColumnString}
                LIMIT ${input.querySize}`,
      };
  }
}

export function getInsert(
  input: SqlAction.TableInput,
  value?: Record<string, any>,
): SqlAction.Output | undefined {
  const label = `Insert`;

  if (!input.columns) {
    return undefined;
  }

  const columns = input.columns;
  const columnString = columns.map((col) => col.name).join(',\n');
  const insertValueString = columns
    .map((col) => {
      if (value?.[col.name] === null) {
        return `null`;
      }
      if (value?.[col.name] !== undefined) {
        // use the value if it's there
        return `'${escapeSQLValue(value[col.name])}'`;
      }
      return `'_${col.name}_'`; // use the default value
    })
    .join(',');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `INSERT INTO ${input.tableId} (${columnString})
                VALUES (${insertValueString})`,
      };
  }
}

// TODO: add a flag to allow keeping the primary key or consistent id
export function getBulkInsert(
  input: SqlAction.TableInput,
  rows?: Record<string, any>[],
): SqlAction.Output | undefined {
  const label = `Insert`;

  if (!input.columns) {
    return undefined;
  }

  if (!rows || rows.length === 0) {
    return undefined;
  }

  const columns = input.columns;
  const columnString = columns.map((col) => col.name).join(',\n');

  const insertValueRows = rows
    ?.map((row) => {
      const cells = columns
        .map((col) => {
          let valToUse = '';
          if (row?.[col.name] !== undefined) {
            // use the value if it's there
            valToUse = `'${escapeSQLValue(row[col.name])}'`;
          } else {
            valToUse = 'null';
          }
          return valToUse;
        })
        .join(',');

      // TODO: see if we need to escape single tick (') for SQL here

      return `(${cells})`;
    })
    .join(', \n');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `INSERT INTO ${input.tableId} (${columnString})
                VALUES ${insertValueRows}`,
      };
  }
}

export function getUpdate(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Update`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = input.columns.map((col) => `${col.name} = ''`).join(',\n');
  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join(' AND \n');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `UPDATE ${input.tableId}
                SET ${columnString}
                WHERE ${whereColumnString}`,
      };
  }
}

export function getUpdateWithValues(
  input: SqlAction.TableInput,
  value: Record<string, any>,
  conditions: Record<string, any>,
): SqlAction.Output | undefined {
  const label = `Update`;

  if (!input.columns) {
    return undefined;
  }

  const columnString = Object.keys(value)
    .map((colName) => {
      let valToUse = value[colName];

      if (!isValueNumber(valToUse)) {
        // wrap the single quote for string
        valToUse = `'${escapeSQLValue(valToUse)}'`;
      }

      return `${colName} = ${valToUse}`;
    })
    .join(', \n');

  const whereColumnString = Object.keys(conditions)
    .map((colName) => {
      let valToUse = conditions[colName];

      if (!isValueNumber(valToUse)) {
        // wrap the single quote for string
        valToUse = `'${escapeSQLValue(valToUse)}'`;
      }

      return `${colName} = ${valToUse}`;
    })
    .join(' AND \n');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `UPDATE ${input.tableId}
                SET ${columnString}
                WHERE ${whereColumnString}`,
      };
  }
}

export function getDelete(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Delete`;

  if (!input.columns) {
    return undefined;
  }

  const whereColumnString = input.columns.map((col) => `${col.name} = ''`).join(' AND \n');

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `DELETE FROM ${input.tableId}
                WHERE ${whereColumnString}`,
      };
  }
}

export function getCreateTable(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Create Table`;

  if (!input.columns) {
    return undefined;
  }

  let columnString: string = '';

  // map the nested column accordingly (using _ as a separator)
  input.columns = input.columns.map((col) => {
    col.name = col.propertyPath?.join('_') || col.name;
    return col;
  });

  // TODO: figure out how to use the defaultval
  switch (input.dialect) {
    case 'mssql':
      columnString = input.columns
        .map((col) =>
          [
            col.name,
            col.type,
            col.primaryKey ? 'PRIMARY KEY' : '',
            col.autoIncrement ? 'IDENTITY' : '',
            col.allowNull ? '' : 'NOT NULL',
          ].join(' '),
        )
        .join(',\n');
      return {
        label,
        formatter,
        query: `CREATE TABLE ${input.tableId} (${columnString})`,
      };
    case 'postgres':
      columnString = input.columns
        .map((col) => {
          const res = [col.name];

          const colType = col.type.toUpperCase();

          // TODO: better use regex here
          // nextval(employees_employeeid_seq::regclass)
          if (
            col.primaryKey &&
            col?.defaultValue?.includes('nextval(') &&
            col?.defaultValue?.includes('_seq::regclass)')
          ) {
            res.push('BIGSERIAL PRIMARY KEY');
          } else {
            if (colType.includes('INT') && col.primaryKey) {
              res.push('BIGSERIAL PRIMARY KEY');
            } else {
              res.push(col.type);
              res.push(col.allowNull ? '' : 'NOT NULL');
            }
          }

          return res.join(' ');
        })
        .join(',\n');
      return {
        label,
        formatter,
        query: `CREATE TABLE ${input.tableId} (${columnString})`,
      };
    case 'sqlite':
      columnString = input.columns
        .map((col) => {
          const colType = col.type.toUpperCase();

          return [
            col.name,
            colType.includes('INT') ? 'INTEGER' : colType,
            col.primaryKey ? 'PRIMARY KEY' : '',
            col.autoIncrement ? 'AUTOINCREMENT' : '',
            col.allowNull ? '' : 'NOT NULL',
          ].join(' ');
        })
        .join(',\n');
      return {
        label,
        formatter,
        query: `CREATE TABLE ${input.tableId} (${columnString})`,
      };
    case 'mariadb':
    case 'mysql':
      columnString = input.columns
        .map((col) =>
          [
            col.name,
            col.type,
            col.primaryKey ? 'PRIMARY KEY' : '',
            col.autoIncrement ? 'AUTO_INCREMENT' : '',
            col.allowNull ? '' : 'NOT NULL',
          ].join(' '),
        )
        .join(',\n');
      return {
        label,
        formatter,
        query: `CREATE TABLE ${input.tableId} (${columnString})`,
      };
  }
}

export function getDropTable(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Drop Table`;

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `DROP TABLE ${input.tableId}`,
      };
  }
}

export function getAddColumn(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Add Column`;

  switch (input.dialect) {
    case 'mssql':
      return {
        label,
        formatter,
        query: `ALTER TABLE ${input.tableId}
                ADD COLUMN newColumn1 NVARCHAR(200)`,
      };
    case 'postgres':
      return {
        label,
        formatter,
        query: `ALTER TABLE ${input.tableId}
                ADD COLUMN newColumn1 CHAR(200)`,
      };
    case 'sqlite':
      return {
        label,
        formatter,
        query: `ALTER TABLE ${input.tableId}
                ADD COLUMN newColumn1 TEXT`,
      };
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `ALTER TABLE ${input.tableId}
                ADD COLUMN newColumn1 varchar(200)`,
      };
  }
}

export function getDropColumns(input: SqlAction.TableInput): SqlAction.Output | undefined {
  const label = `Drop Column`;

  if (!input.columns) {
    return undefined;
  }

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: input.columns
          .map(
            (col) => `ALTER TABLE ${input.tableId}
                         DROP COLUMN ${col.name};`,
          )
          .join('\n'),
      };
  }
}

export function getDropDatabase(input: SqlAction.DatabaseInput): SqlAction.Output | undefined {
  const label = `Drop Database`;

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `DROP DATABASE ${input.databaseId}`,
      };
  }
}

export function getCreateDatabase(input: SqlAction.DatabaseInput): SqlAction.Output | undefined {
  const label = `Create Database`;

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `CREATE DATABASE ${input.databaseId}`,
      };
  }
}

export function getCreateSampleTable(input: SqlAction.DatabaseInput): SqlAction.Output | undefined {
  const label = `Create Table`;

  let query = '';

  switch (input.dialect) {
    case 'mssql':
      query = `
        CREATE TABLE mocked_table
        (
          id INTEGER PRIMARY KEY IDENTITY NOT NULL,
          name NVARCHAR(120)
        );
      `;
      break;
    case 'postgres':
      query = `
        CREATE TABLE mocked_table
        (
          id BIGSERIAL PRIMARY KEY,
          name CHAR(120)
        );
      `;
      break;
    case 'sqlite':
      query = `
        CREATE TABLE mocked_table (
          id INTEGER PRIMARY KEY NOT NULL,
          name NVARCHAR(120)
        )
      `;
      break;
    case 'mariadb':
    case 'mysql':
      query = `
        CREATE TABLE mocked_table
        (
          id INTEGER PRIMARY KEY AUTO_INCREMENT NOT NULL,
          name NVARCHAR(120)
        );
      `;
      break;
  }

  return {
    label,
    formatter,
    query,
    skipGuide: true,
  };
}

export function getCreateConnectionDatabase(
  input: SqlAction.ConnectionInput,
): SqlAction.Output | undefined {
  const label = `Create Database`;

  switch (input.dialect) {
    case 'mssql':
    case 'postgres':
    case 'sqlite':
    case 'mariadb':
    case 'mysql':
      return {
        label,
        formatter,
        query: `CREATE DATABASE some_database_name`,
      };
  }
}

export class ConcreteDataScripts extends BaseDataScript {
  dialects = ['mysql', 'mariadb', 'mssql', 'postgres', 'sqlite'];

  getIsTableIdRequiredForQuery() {
    return false;
  }

  getSyntaxMode() {
    return 'sql';
  }

  supportMigration() {
    return true;
  }

  supportCreateRecordForm() {
    return true;
  }

  supportEditRecordForm() {
    return true;
  }

  getTableScripts() {
    return [
      getSelectAllColumns,
      getSelectCount,
      getSelectSpecificColumns,
      getSelectDistinctValues,
      getDivider,
      getInsert,
      getUpdate,
      getDelete,
      getDivider,
      getCreateTable,
      getDropTable,
      getAddColumn,
      getDropColumns,
    ];
  }

  getDatabaseScripts() {
    return [getDivider, getDropDatabase, getCreateDatabase, getDivider, getCreateSampleTable];
  }

  getConnectionScripts() {
    return [getDivider, getCreateConnectionDatabase];
  }

  getSampleConnectionString(dialect) {
    switch (dialect) {
      case 'mssql':
        return `mssql://sa:password123!@localhost:1433`;
      case 'postgres':
        return `postgres://postgres:password@localhost:5432`;
      case 'sqlite':
        return `sqlite://test-db.sqlite`;
      case 'mariadb':
        return `mariadb://root:password@localhost:3306`;
      case 'mysql':
        return `mysql://root:password@localhost:3306`;
      default: // Not supported dialect
        return '';
    }
  }

  getSampleSelectQuery(actionInput: SqlAction.TableInput) {
    return getSelectAllColumns(actionInput);
  }

  // sample code snippet
  getCodeSnippet(connection: string, database: string, language: SqluiCore.LanguageMode, sql: string) {
    switch (language) {
      case 'javascript':
        return `
// install these extra dependencies if needed
// npm install --save sequelize
// npm install --save pg pg-hstore # optional - Postgres
// npm install --save mysql2 # optional
// npm install --save mariadb # optional
// npm install --save sqlite3 # optional
// npm install --save tedious # optional - Microsoft SQL Server
const {Sequelize} = require('sequelize');

async function _doWork(){
  const sequelize = new Sequelize('${connection}');

  try{
    const [items, meta] = await sequelize.query(\`${sql}\`, {
      raw: true,
      plain: false,
    });

    for(const item of items){
      console.log(item);
    }
  } catch(err){
    console.log('Failed to run query', err);
  }
}

_doWork();
        `.trim();
      case 'python':
        // NOTE: for sqlite, sqlalchemy needs an extra /
        connection = connection.replace('sqlite://', 'sqlite:///')
          .replace('postgres://', 'postgresql://')//SQLAlchemy used to accept both, but has removed support for the postgres name - https://stackoverflow.com/questions/62688256/sqlalchemy-exc-nosuchmoduleerror-cant-load-plugin-sqlalchemy-dialectspostgre

        // if there is a database, then append it
        if (database){
          connection += `/${database}`;
        }

        return `
# python3 -m venv ./ # setting up virtual environment with
# source bin/activate # activate the venv profile
# pip install sqlalchemy
# pip install mysqlclient # optional - for MySQL
# pip install psycopg2-binary # optional - for Postgres
# pip install pyodbc # optional - for MSSQL

from sqlalchemy import create_engine
from sqlalchemy.sql import text

engine = create_engine('${connection}', echo = True)

with engine.connect() as con:
  rs = con.execute("""${sql}""")

  for row in rs:
    print(row)
        `.trim();
      default:
        return '';
    }
  }
}

export default new ConcreteDataScripts();
