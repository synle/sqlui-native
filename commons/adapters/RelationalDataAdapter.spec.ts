import RelationalDataAdapter from './RelationalDataAdapter';

const adapter = new RelationalDataAdapter('sqlite://mocked-db.sqlite');

describe('sqlite', () => {
  test('sqlite - Create and insert table', async () => {
    try {
      // create the table
      await adapter.execute(`
      CREATE TABLE artists (
        ArtistId INTEGER PRIMARY KEY NOT NULL,
        Name NVARCHAR(120)
      )
    `);

      // insert some record
      await adapter.execute(`
      INSERT INTO artists (Name) VALUES ('Test Artist 1');
    `);
      await adapter.execute(`
      INSERT INTO artists (Name) VALUES ('Test Artist 2');
    `);
      await adapter.execute(`
      INSERT INTO artists (Name) VALUES ('Test Artist 3');
    `);

      expect(1).toBe(1);
    } catch (err) {
      expect(err).toBeUndefined();
    }
  });

  test('sqlite - Get tables', async () => {
    const tables = await adapter.getTables();
    expect(tables).toMatchSnapshot();
  });

  test('sqlite - Get columns', async () => {
    const columns = await adapter.getColumns('artists');
    expect(columns).toMatchSnapshot();
  });

  test('sqlite - Execute Select', async () => {
    const resp = await adapter.execute(`SELECT * FROM artists ORDER BY Name ASC LIMIT 10`);
    expect(resp?.raw?.length).toBe(3);
    expect(resp).toMatchSnapshot();
  });

  test('sqlite - Execute Update', async () => {
    try{
    const resp = await adapter.execute(`UPDATE artists SET name = 'AC/DC' WHERE ArtistId = '1'`);
    expect(1).toBe(1);
    } catch (err) {
      expect(err).toBeUndefined();
    }
  });
});
