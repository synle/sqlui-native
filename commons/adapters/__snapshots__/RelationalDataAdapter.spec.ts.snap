// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqlite sqlite - Execute Select 1`] = `
Object {
  "affectedRows": undefined,
  "meta": undefined,
  "ok": true,
  "raw": Array [
    Object {
      "ArtistId": 1,
      "Name": "Test Artist 1",
    },
    Object {
      "ArtistId": 2,
      "Name": "Test Artist 2",
    },
    Object {
      "ArtistId": 3,
      "Name": "Test Artist 3",
    },
  ],
}
`;

exports[`sqlite sqlite - Get columns 1`] = `
Array [
  Object {
    "allowNull": false,
    "defaultValue": undefined,
    "name": "ArtistId",
    "primaryKey": true,
    "type": "INTEGER",
    "unique": false,
  },
  Object {
    "allowNull": true,
    "defaultValue": undefined,
    "name": "Name",
    "primaryKey": false,
    "type": "NVARCHAR(120)",
    "unique": false,
  },
]
`;

exports[`sqlite sqlite - Get tables 1`] = `
Array [
  Object {
    "columns": Array [],
    "name": "artists",
  },
]
`;
